tools/internal/shell/real.go

```diff
@@ -0,0 +1,39 @@
+package shell
+
+import (
+	"github.com/rs/zerolog/log"
+	"os"
+	"os/exec"
+)
+
+// RealRunner is an implementation of the Runner interface that actually executes shell commands
+type RealRunner struct{}
+
+// NewRealRunner constructs a new RealRunner
+func NewRealRunner() *RealRunner {
+	return &RealRunner{}
+}
+
+// Run runs a Command, returning an error if the command exits non-zero
+func (r *RealRunner) Run(cmd Command) error {
+	execCmd := exec.Command(cmd.Prog, cmd.Args...)
+	execCmd.Dir = cmd.Dir
+
+	if !cmd.PristineEnv {
+		execCmd.Env = os.Environ()
+	}
+	execCmd.Env = append(execCmd.Env, cmd.Env...)
+
+	// TODO - would be nice to capture out/err and stream to debug log, to cut down on noise
+	execCmd.Stdout = os.Stdout
+	execCmd.Stderr = os.Stderr
+
+	log.Info().Msgf("Executing: %q", cmd.PrettyFormat())
+	err := execCmd.Run()
+	if err != nil {
+		log.Error().Msgf("Command failed: %v", err)
+		return &Error{Command: cmd, Err: err}
```

Comment:
can you return `&Error` when Run expects to return `error`?

Reply:
I think there may be some confusion over what go's `error` type actually is here (correct me if I'm wrong). `error is a single method interface that looks like: 
```
type interface error {
    Error() string
}
```

So anything that has a method `Error()` matching that signature including pointers to structs can be the `error` type

Reply:
I mean, it was my choice to implement Error with a pointer receiver and not a value receiver. :) Which I did solely because of examples I found online.

This SO post has more detail about why to use pointers and not values for custom errors (https://stackoverflow.com/questions/50333428/custom-errors-in-golang-and-pointer-receivers), seems to boil down to "so you don't break identity comparisons for errors".

Reply:
Yeah I think in general it's pretty common to implement interfaces on pointer receivers :) Particularly if the methods need to mutate the internal state of the receiver object.